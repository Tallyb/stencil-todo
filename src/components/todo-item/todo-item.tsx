import { Component, Prop, Event, EventEmitter, State } from '@stencil/core';

@Component({
  tag: 'todo-item',
  styleUrl: 'todo-item.css'
})

export class TodoItem {
  
  @Prop() name: string; 
  @Prop() done: boolean;
  @Prop() index: number;

  @State() editMode = false;

  @Event() itemChanged: EventEmitter<any>;
  @Event() itemRemoved: EventEmitter<number>;
  @Event() itemToggled: EventEmitter<number>;

  emitItem() {
      this.itemChanged.emit({
        index: this.index,
        name: this.name,
        done: this.done
      });
  }
  onInputChanged(ev) {
    console.log('Value', ev.value)
    this.itemChanged.emit({index: this.index, name: ev.value});
  }

  toggleCompletion() {
    this.itemToggled.emit(this.index);    
  }

  remove() {
    this.itemRemoved.emit(this.index);
  }
  edit () {

  }
  render() {
    return (
      <li>
        <div class="view">
            <input class="toggle" type="checkbox" onClick={() => this.toggleCompletion()} checked={this.done}/>
            <label onClick={() => this.edit()}>{this.name}</label>
            <button class="destroy" onClick={() =>this.remove()}></button>
		</div>
        <input class="edit" type="text"
            placeholder="New Todo"
            value={this.name}
            onInput={() => this.onInputChanged}
        />
      </li>
    );
  }
}