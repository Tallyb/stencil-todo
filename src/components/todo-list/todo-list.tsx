import { Component, State, Event, EventEmitter, Prop } from '@stencil/core';
import { ItemsService } from '../../services/items';

@Component({
  tag: 'todo-list',
  styleUrl: 'todo-list.css'
})
export class TodoList {

    @Event() todoItemsChanged: EventEmitter<any>;
    
    @Prop({mutable: true}) items = []; 

    private itemsService = new ItemsService(); 

    async componentWillLoad (){
        if(!this.items) {
            this.items =  await this.itemsService.get()
        }
    }

    changeItem(index, value) {
        this.items = [...this.items.slice(0, index),
        {
            name: value, 
            done: false
        },
        ... this.items.slice(index + 1)];
        this.todoItemsChanged.emit(this.items);
    }

    removeItem(index) {
        this.items = [
            ...this.items.slice(0, index),
            ...this.items.slice(index + 1)
        ];
        this.todoItemsChanged.emit(this.items);
    }

    toggleItem(index) {
        this.items = [
            ...this.items.slice(0, index),
            {
                name: this.items[index].name, 
                done: !this.items[index].done
            },
            ...this.items.slice(index + 1)
        ];
        this.todoItemsChanged.emit(this.items);
    }

    getRemainingItems() {
        return this.items.filter( i => !i.done).length;
    }

    addItem(ev) {
        if (ev.key == 'Enter' && ev.target.value) {
            this.items = [{name: ev.target.value, done: false}, ...this.items ];
            ev.target.value = '';
        }        
    }

    renderMain(): JSX.Element {
        return (
            <section class="main">
                <ul class="todo-list">
                    {this.items.map((i, k) =>
                        <todo-item name={i.name} done={i.done}
                            onItemChanged={ ev => this.changeItem(k, ev.detail)}
                            onItemRemoved={() => this.removeItem(k)}
                            onItemToggled={() => this.toggleItem(k)}
                        ></todo-item>
                    )}
                </ul>
            </section>
        )
    }

    render() {
        return (
          <div>
            <input class="new-todo" 
                placeholder="What needs to be done?" 
                autofocus=""
                onKeyPress={ev => this.addItem(ev)}
            />
            {this.items.length === 0 ? '': this.renderMain()}
            <footer class="footer">
                <span class="todo-count"><strong> {this.getRemainingItems()} Left</strong> </span>
            </footer>
          </div>
    );
    }
}

