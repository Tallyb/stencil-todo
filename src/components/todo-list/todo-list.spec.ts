import { TestWindow } from '@stencil/core/testing';
import { TodoList } from './todo-list';
import { TodoItem } from '../todo-item/todo-item';

describe('todo-list', () => {
    let win;
    beforeEach(() => {
        win = new TestWindow();
    });

    const ITEMS = [
        { name: 'Item 1', done: false},
        { name: 'Item 2', done: false},
        { name: 'Item 3', done: true}
    ];

    describe('remaining items', () => {
        it('should display remaining items', () => {
            let comp = new TodoList();
            comp.items = [
                { name: 'an item', done: true}
            ]
            expect(comp.getRemainingItems()).toEqual(0);
        });
    
        it('should display remaining items', () => {
            let comp = new TodoList();
            comp.items =ITEMS;
            expect(comp.getRemainingItems()).toEqual(2);
        });        
        
    });

    describe('events payload', () => {
        it('should remove item from list on remove', () => {
    
            let comp = new TodoList();
            comp.items = ITEMS;
            const changedSpy = jest.fn();
            comp.todoItemsChanged = {
                emit: changedSpy
            };
            //const changedSpy = jest.spyOn(comp.todoItemsChanged, 'emit');
            comp.removeItem(1);
            expect(changedSpy).toHaveBeenCalledWith([
                { name: 'Item 1', done: false},
                { name: 'Item 3', done: true}
            ]);


        });
    });

    describe('rendering', () => {
        it('should render with empty list', async () => {
            let element = await win.load({
                components: [TodoList],
                html: '<todo-list></todo-list>'
            }); 
            expect(element).toMatchSnapshot();
        });
        
        it('should render with items list', async () => {
            let element = await win.load({
                components: [TodoList],
                html: '<todo-list></todo-list>'
            });
            element.items = [
                {name: 'Milk the cow', done: true},
                {name: 'Buy milk', done: false}
            ];
            win.flush(); 
            expect(element).toMatchSnapshot();
        });        
    });

    describe('items editing', () => {
        let element;
        beforeEach( async () => {
            element = await win.load({
                components: [TodoList, TodoItem],
                html: '<todo-list></todo-list>'
            });
            element.items = ITEMS;
            win.flush();
        });
        it('should toggle second item', async () => {
            let eventSpy = jest.fn(); 
            win.document.addEventListener('todoItemsChanged', eventSpy);
            let toggles = element.querySelectorAll('.toggle');
            toggles[1].click();
            win.flush();
            expect(eventSpy).toHaveBeenCalled();
            expect(eventSpy.mock.calls[0][0].detail[2].done).toEqual(true);
        });

        fit('should update second item', async () => {
            let eventSpy = jest.fn(); 
            win.document.addEventListener('todoItemsChanged', eventSpy);
            let todo = element.querySelectorAll('todo-item')[1];
            todo.querySelector('label').click();
            win.flush();
            // create an input event
            let change = win.document.createEvent('Event');
            change.initEvent('change', true, false);
            // Set value on the input element & dispatch the event
            let value = 'New Value';
            let input = element.querySelector('.edit');
            input.value = value;
            input.dispatchEvent(change);
            win.flush();
            // expectations
            expect(eventSpy).toHaveBeenCalled();
            expect(eventSpy.mock.calls[0][0].detail[1].name).toEqual(value);
        });
        
    });
});