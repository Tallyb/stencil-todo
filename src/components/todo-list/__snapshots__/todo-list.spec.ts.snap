// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo-list rendering should do deep rendering 1`] = `
<todo-list
  class="hydrated"
>
  <!---->
  <div>
    <input
      class="new-todo"
      placeholder="What needs to be done?"
    />
    <section
      class="main"
    >
      <ul
        class="todo-list"
      >
        <todo-item
          class="hydrated"
        >
          <!---->
          <li>
            <div
              class="view"
            >
              <input
                class="toggle"
                type="checkbox"
              />
              <label>
                Item 1
              </label>
              <button
                class="destroy"
              />
            </div>
          </li>
        </todo-item>
        <todo-item
          class="hydrated"
        >
          <!---->
          <li>
            <div
              class="view"
            >
              <input
                class="toggle"
                type="checkbox"
              />
              <label>
                Item 2
              </label>
              <button
                class="destroy"
              />
            </div>
          </li>
        </todo-item>
        <todo-item
          class="hydrated"
        >
          <!---->
          <li>
            <div
              class="view"
            >
              <input
                class="toggle"
                type="checkbox"
              />
              <label>
                Item 3
              </label>
              <button
                class="destroy"
              />
            </div>
          </li>
        </todo-item>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        <strong>
           2 Left
        </strong>
         
      </span>
    </footer>
  </div>
</todo-list>
`;

exports[`todo-list rendering should do shallow rendering 1`] = `
<todo-list
  class="hydrated"
>
  <!---->
  <div>
    <input
      class="new-todo"
      placeholder="What needs to be done?"
    />
    <section
      class="main"
    >
      <ul
        class="todo-list"
      >
        <todo-item
          done="false"
          name="Item 1"
        />
        <todo-item
          done="false"
          name="Item 2"
        />
        <todo-item
          done="true"
          name="Item 3"
        />
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        <strong>
           2 Left
        </strong>
         
      </span>
    </footer>
  </div>
</todo-list>
`;
